src/hyperloglog.rs

```diff
@@ -363,91 +364,200 @@ where
-///
-/// # Examples
-/// ```
-/// use pdatastructs::hyperloglog::HyperLogLog;
-/// use pdatastructs::hyperloglog::serde::HyperLogLogExport;
-/// use serde_json;
-/// use std::hash::BuildHasherDefault;
-/// use std::collections::hash_map::DefaultHasher;
-///
-/// let mut hll = HyperLogLog::new(4);
-/// hll.add(&"some string");
-///
-/// let hll_export = HyperLogLogExport::from(&hll);
-/// let json = serde_json::to_string(&hll_export).expect("valid json");
-///
-/// // elsewhere
-/// let hll_export = serde_json::from_str(&json).expect("valid export");
-/// let hll: HyperLogLog<str, BuildHasherDefault<DefaultHasher>> = HyperLogLog::from_exported_hyperloglog(hll_export, BuildHasherDefault::default());
-///
-/// assert!(hll.count() > 0);
-/// ```
 #[cfg(feature = "serde")]
 pub mod serde {
-
     use super::HyperLogLog;
+    use serde::de::{self, Visitor};
+    use serde::ser::SerializeStruct;
     use serde::{Deserialize, Serialize};
     use std::fmt;
     use std::hash::{BuildHasher, Hash};
     use std::marker::PhantomData;
 
-    impl<T, B> HyperLogLog<T, B>
+    impl<T, B> Serialize for HyperLogLog<T, B>
     where
         T: Hash + ?Sized,
-        B: BuildHasher + Clone + Eq,
+        B: BuildHasher + Clone + Eq + Serialize,
     {
-        /// Reconstruct HyperLogLog from a previously exported HyperLogLog
-        pub fn from_exported_hyperloglog(exported: HyperLogLogExport, buildhasher: B) -> Self {
-            Self {
-                registers: exported.registers,
-                b: exported.b,
-                buildhasher,
-                phantom: PhantomData,
-            }
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: serde::Serializer,
+        {
+            let mut state = serializer.serialize_struct("HyperLogLog", 3)?;
+            state.serialize_field("registers", &self.registers)?;
+            state.serialize_field("b", &self.b)?;
+            state.serialize_field("buildhasher", &self.buildhasher)?;
+            state.end()
         }
     }
 
-    /// Captures the underlying HyperLogLog data such that it can be serialized and deserialized
-    /// using serde
-    #[derive(Clone, Serialize, Deserialize, Eq, PartialEq)]
-    pub struct HyperLogLogExport {
-        /// HLL Data
-        pub registers: Vec<u8>,
-        /// Number of bits used for register selection
-        pub b: usize,
-    }
-
-    impl<T, H> From<HyperLogLog<T, H>> for HyperLogLogExport
+    impl<'de, T, B> Deserialize<'de> for HyperLogLog<T, B>
     where
         T: Hash + ?Sized,
-        H: BuildHasher + Clone + Eq,
+        B: BuildHasher + Clone + Eq + Deserialize<'de>,
     {
-        fn from(value: HyperLogLog<T, H>) -> Self {
-            Self {
-                registers: value.registers,
-                b: value.b,
+        fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+        where
+            D: serde::Deserializer<'de>,
+        {
+            enum Field {
```

Comment:
last nitpick: can you move the structs and impls into the `hyperloglog::serde` module? That makes the code a bit easier to read and the `serde` module prevents that stuff from leaking all over the place.

Reply:
cool, I've moved the hll and data to a module and extracted serde. I've added a public getter for the registers, because as I said, having just a direct serde impl doesn't actually help for my use case, which is kind of why I started the PR in the first place.